### ISSUE TITLE
macOS Distribution

### ISSUE THREAD

**Issue by lorenzolewis** [tauri-apps/tauri#4415]
Sparked by https://github.com/tauri-apps/tauri-docs/issues/691, related to https://github.com/tauri-apps/tauri/issues/1731 and https://github.com/tauri-apps/tauri/issues/3055

I'm mainly using this issue to keep track of my thoughts because this is going to be a long ride...

## Package Creation

You must have your app in a .ipa (iPad, iOS, some cross-compiled macOS) or a .pkg (macOS) format prior to upload. There's a reference to a .dmg, but I'm not 100% sure if this is supported or not.

- Need to know the best way to achieve this
- Currently wondering if `xcodebuild -exportArchive` achieves this

## Upload Application to App Store Connect

Ways to upload:

- `xcrun notarytool`, I believe this is the current preferred way for Xcode 13 and later (https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution/customizing_the_notarization_workflow#3087734)
- `xcrun altool` I believe this is what to use prior to Xcode 13 (https://developer.apple.com/forums/thread/64041)
- Transporter, GUI app in the Mac App Store (https://apps.apple.com/gb/app/transporter/id1450874784?mt=12)

---

## Additional References

https://developer.apple.com/documentation/security/notarizing_macos_software_before_distribution/customizing_the_notarization_workflow/customizing_the_xcode_archive_process

CC @nothingismagick @lucasfernog

**Comment by lorenzolewis** [comment]
Also, with these things in mind, if we really do want to bundle in automatic notarisation and bundling, I don't know how easily we can compile cross-platform given this all HEAVILY depends on Xcode utils

**Comment by lorenzolewis** [comment]
@lucasfernog how does this all fit into the codebase now? Is `tauri-core`/`bundler` responsible for the pkg/dmg creation? Then our current notarisation flow is handled via our GitHub action?

**Comment by lucasfernog** [comment]
tauri-bundler is responsible for bundling `.app` and `.dmg` for macOS and signing/notarizing it if code sign is setup @lorenzolewis

**Comment by lucasfernog** [comment]
We don't have .pkg yet.

**Comment by lucasfernog** [comment]
> Also, with these things in mind, if we really do want to bundle in automatic notarisation and bundling, I don't know how easily we can compile cross-platform given this all HEAVILY depends on Xcode utils

We do have automatic notarization and signing already using the xcode tools (though we could check https://github.com/indygreg/PyOxidizer/tree/main/apple-codesign in the future).

**Comment by bencevans** [comment]
This can be done manually using Xcode's `productbuild` CLI, can take a .app and export a .pkg <sup>[1]</sup>

[1]: https://www.miradore.com/knowledge/macos/converting-dmg-to-pkg-for-macos-application-deployment/

**Comment by buildnewapp** [comment]
how can i signing/notarizing app through cli?

**Comment by ryan-bunkr** [comment]
@buildnewapp this is my current approach:
first, use the tauri cli to build the app to a .dmg and extract the .app locally (build with an "Apple Distribution" certificate)

then, with a "Mac Installer Distribution" certificate in your local mac keychain, run product build on the .app file like this:

`productbuild --component YOUR_APP.app /Applications --sign "3rd Party Mac Developer Installer: YOUR COMPANY (XXXXXX)" YOUR_APP.pkg`

**Comment by rendaoer** [comment]
> @buildnewapp this is my current approach: first, use the tauri cli to build the app to a .dmg and extract the .app locally (build with an "Apple Distribution" certificate)
> 
> then, with a "Mac Installer Distribution" certificate in your local mac keychain, run product build on the .app file like this:
> 
> `productbuild --component YOUR_APP.app /Applications --sign "3rd Party Mac Developer Installer: YOUR COMPANY (XXXXXX)" YOUR_APP.pkg`

I can't use the Apple distribution certificate to package [pnpm tauri build]. How did you manage to package it successfully?

**Comment by rendaoer** [comment]
Thanks guys, I've successfully uploaded TestFlight!!!

**Comment by paprikka** [comment]
> Thanks guys, I've successfully uploaded TestFlight!!!

@rendaoer Did you have to create a provisioning profile for your app? How did you managed to get this to work? Are you using Transporter? I feel like there's a small detail missing in my config.

FYI this is the script I'm using to build the final package and push it to AppStore Connect using Transporter:

### Requirements: 

- entitlements.plist in the project root dir (referenced in tauri.conf)

### Code:

```bash
PKG_SIGNING_IDENTITY="3rd Party Mac Developer Installer: NAME (ID)"
PROVISIONING_PROFILE="./embedded.provisionprofile" # to enable TestFlight builds
BUNDLE_NAME=Enso # name w/o the extension
SRC_BUNDLE_PATH="./src-tauri/target/universal-apple-darwin/release/bundle/macos/$BUNDLE_NAME.app"

rm -rf build
mkdir build
npm run tauri build -- --target universal-apple-darwin

cp $PROVISIONING_PROFILE "$SRC_BUNDLE_PATH/Contents/" 

productbuild --component $SRC_BUNDLE_PATH "/Applications/$BUNDLE_NAME.app" ./build/unsigned.pkg

productsign --sign $PKG_SIGNING_IDENTITY ./build/unsigned.pkg ./build/signed.pkg

open ./build
```

### Notes:

1) No need to manually extract the `AppName.app` file from `.dmg` as in some earlier comments
2) All looks good when I verify the build in the Transporter App, _but_ _invalid signature_ and _missing entitlements_ errors are returned from AppStore Connect after the upload. This is a bit annoying because I addressed these issues when they appeared in the Transporter app verification at first. IIRC once you hit "verify" in the Transporter app, verification is done via the AppStore Connect API... 🤷🏽‍♀️. 

Hope the snippet is useful. If you have an idea how to tackle 2), let me know. I'm running out of ideas.

**Comment by rendaoer** [comment]
@paprikka 
Show you my shell and you'll understand
```shell
targets=("aarch64-apple-darwin" "x86_64-apple-darwin" "universal-apple-darwin");

target=${targets[$1]};

app_path="src-tauri/target/${target}/release/bundle/macOS/Name.app";

sign_app="3rd Party Mac Developer Application: Inc Name Co., Ltd. (key)";

sign_install="3rd Party Mac Developer Installer: Inc Name Co., Ltd. (key)";

build_name="Name.pkg";

profile="embedded.provisionprofile";

cp_dir="src-tauri/target/${target}/release/bundle/macOS/Name.app/Contents/";

pnpm tauri build --target "${target}" --verbose;
cp "${profile}" "${cp_dir}";

codesign --force -s "${sign_app}" --entitlements Name.entitlements "${app_path}";

productbuild --component "${app_path}" /Applications/ --sign "${sign_install}" "${build_name}";
```
That's about it, no environment variables are used to sign. The configuration and authorization are written manually, then signed and packaged.

**Comment by paprikka** [comment]
Thanks @rendaoer, the script worked 👌🏼.

In case someone else makes the same mistake as I have: the issue was that I was using the _regular distribution_ profile (`Apple Distribution: Company (key)`) to sign the app, instead of a _3rd Party Developer_ profile (`3rd Party Mac Developer Application: Inc Name Co., Ltd. (key)`).

Another difference was that I was using `productsign` instead of `codesign`. However, this should be irrelevant as productsign is just a subset of what codesign does ([src](https://stackoverflow.com/questions/22385320/whats-the-difference-between-codesign-and-productsign)).

**Comment by paprikka** [comment]
Edit: in my case the build, signing and validation using (Transporter) work like charm.

However, once I push the TF build to AppStore connect, I get the dreaded "not available for testing" message after a few hours of validation. 

I'm using the certs mentioned in the post above (`XXX "3rd Party Mac Developer Application: Company (Key)"`, `XXX "3rd Party Mac Developer Installer: Company (Key)"`) and a distribution provisioning profile connected to the former.

**Comment by edjw** [comment]
Where are you all getting the `embedded.provisionprofile` file from? I can't find that anywhere.

Thanks to all for the help on this thread

**Comment by paprikka** [comment]
I managed to push the app to AppStore connect and enable private TestFlight tests. 
The issue was my entitlements.plist file. It seems like we need to add a few more fields:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>

    <key>com.apple.security.app-sandbox</key><true/>

    <key>com.apple.application-identifier</key>
    <string>{TEAM_ID}.{BUNDLE_ID}</string>
    
    <key>com.apple.developer.team-identifier</key>
    <string>{TEAM_ID}</string>
    
    <!-- Missing in my case, but probably irrelevant in yours -->
    <key>com.apple.security.files.user-selected.read-write</key>
    <true/>

  </dict>
</plist>
 
```

Now, there seems to be another issue at play here as I can't create a public TestFlight release and just before submitting it for review we get served a screen saying that this app was created using a beta version of Xcode.

<img width="715" alt="image" src="https://user-images.githubusercontent.com/1511643/207693554-97305bcd-18aa-426c-a8a1-3f066701aba3.png">

I'm on Xcode 14.2 (stable), so I suspect that this is due to missing metadata or a heuristic, since I copied the entitlements for an app built using Xcode to fix the problem mentioned in the earlier comments.

A potential hack/fix could be adding some custom fields to the plist file included in the bundle to essentially tell the AppStore that we're Xcode. 

Would it be possible for you to share: the structure of your entitlements file, your version of Xcode and whether/how you're customising the info.plist file (@rendaoer  et al)?


TL;DR if the scripts above work but you end up seeing this in your build list:

<img width="538" alt="image" src="https://user-images.githubusercontent.com/1511643/207696310-387c1abc-f069-431b-bafd-9243402fc22d.png">

it's possible that you need to update your entitlements (altool --validate or Transporter won't detect the issue).

EDIT:

After installing the app via TestFlight (private, public still doesn't work), I get a message saying "The application APP_NAME can't be opened". No explanation, back to square one. I'm gonna stop commenting to avoid making this thread more confusing.

**Comment by paprikka** [comment]
@edjw I posted an example entitlements.plist file above. I keep it in the project root dir and reference in tauri.conf like this:

```json
"macOS": {
        "entitlements": "../entitlements.plist",
        "frameworks": []
      },
```

And I get it from: https://developer.apple.com/account/resources/profiles/list

Remember to attach it to the right certificate.

Useful links:
https://developer.apple.com/forums/thread/701514#701514021
https://developer.apple.com/forums/thread/701514#701514021

**Comment by prenx4x** [comment]
After hours of debugging and research I finally made it work and successfully submitted my app to the app store.
Thanks @paprikka  and @rendaoer for your inputs.

The error with testflight was due to missing metadata as pointed out by @paprikka. Specifically 
```
<key>DTXcode</key>
<string>1331</string>
<key>DTXcodeBuild</key>
<string>13E500a</string>
```

I had to add those manually, after which testflight accepted my build. This seems like a bug with tauri builder and it should add such fields automatically.

I also modified the build script to create 2 version of the app - one signed with dev cert and another with dist cert. This will allow to test the signed app locally.

```
targets=("aarch64-apple-darwin" "x86_64-apple-darwin" "universal-apple-darwin");

target=${targets[$1]};

app_path="src-tauri/target/${target}/release/bundle/macOS/MyApp.app";
app_path_dev="src-tauri/target/${target}/release/bundle/macOS/MyApp-dev.app";

sign_app="Apple Distribution: My Company (XXXXXXXXX)";
sign_app_dev="Apple Development: My Cert (XXXXXXX)";

sign_install="3rd Party Mac Developer Installer: My Company (XXXXXXXX)";

build_name="src-tauri/target/${target}/release/bundle/macOS/MyApp.pkg";

profile="src-tauri/entitlements/embedded.provisionprofile";
profile_dev="src-tauri/entitlements/dev/embedded.provisionprofile";

cp_dir="src-tauri/target/${target}/release/bundle/macOS/MyApp.app/Contents/";
cp_dir_dev="src-tauri/target/${target}/release/bundle/macOS/MyApp-dev.app/Contents/";

npm run build:mac;
cp -r "${app_path}" "${app_path_dev}"
cp "${profile}" "${cp_dir}";
cp "${profile_dev}" "${cp_dir_dev}";

codesign --force -s "${sign_app}" --entitlements ./src-tauri/entitlements/entitlements.mac.plist "${app_path}";
codesign --force -s "${sign_app_dev}" --entitlements ./src-tauri/entitlements/entitlements.mac.plist "${app_path_dev}";

productbuild --component "${app_path}" /Applications/ --sign "${sign_install}" "${build_name}";
```

Also if your app is showing a blank window after codesigning you may be missing some entitlements needed by your app.
Tauri docs needs to mention clearly that it does not build pkg needed for app store submission and steps needed to build the same. But long term need better support building for app stores for both windows and mac os.

**Comment by wtto00** [comment]
After upgrading to `tauri v1.5`, I finally managed to successfully package it for `TestFlight`. I tested it and it can be downloaded and installed from `TestFlight`, and runs well.
Thank you all for the examples above.
![image](https://github.com/tauri-apps/tauri/assets/30424139/815fe774-1d37-4c80-b6e4-80b131a9f15d)

Here is my script, hopefully it can help some people:

```shell
export APPLE_SIGNING_IDENTITY="Developer ID Application: my company name (XXXXXXXX)" # replace my company name (XXXXXXXX)
export APPLE_CERTIFICATE_PASSWORD="my certificate password" # replace my certificate password
export APPLE_CERTIFICATE=$(cat "path/to/certificate-base64.txt") # replace path/to/certificate-base64.txt
export APPLE_API_ISSUER="my apple account api issue" # replace my apple account api issue
export APPLE_API_KEY="my apple account api key" # replace my apple account api key
export APPLE_API_KEY_PATH="path/to/AuthKey_key.p8" # replace path/to/AuthKey_key.p8

target="universal-apple-darwin"

# build & first code sign
npx tauri build --target "${target}" --verbose

app_path="src-tauri/target/${target}/release/bundle/macos/my app name.app" # replace my app name.app

# copy profile
profile="path/to/Mac_App_Distribution.provisionprofile" # replace path/to/Mac_App_Distribution.provisionprofile
profile_dest_dir="${app_path}/Contents/embedded.provisionprofile"
cp "${profile}" "${profile_dest_dir}"

# codesign with 3rd Party Mac Developer Application
sign_app="3rd Party Mac Developer Application: my company name (XXXXXXXX)"
entitlements="path/to/app.entitlements" # replace path/to/app.entitlements
codesign --deep --force -s "${sign_app}" --entitlements ${entitlements} "${app_path}"

# package pkg file
pkg_name="my app name.pkg" # replace my app name.pkg
sign_install="3rd Party Mac Developer Installer: my company name (XXXXXXXX)" # replace my company name (XXXXXXXX)
productbuild --component "${app_path}" /Applications/ --sign "${sign_install}" "${pkg_name}"
```

- Where `path/to/certificate-base64.txt` comes from the `p12` file exported from the certificate `Developer ID Application: my company name (XXXXXXXX)`, execute `openssl base64 -in /path/to/certificate.p12 -out certificate-base64.txt`.
- The certificate `3rd Party Mac Developer Application: my company name (XXXXXXXX)` comes from selecting `Mac App Distribution` in https://developer.apple.com/account/resources/certificates/add. Download the certificate and install it on your machine.
- The certificate `3rd Party Mac Developer Installer: my company name (XXXXXXXX)` comes from selecting `Mac Installer Distribution` in https://developer.apple.com/account/resources/certificates/add. Download the certificate and install it on your machine.
- The `path/to/Mac_App_Distribution.provisionprofile` is obtained from selecting `Mac App Store` in `https://developer.apple.com/account/resources/profiles/add` and downloading it locally.
- My `path/to/app.entitlements` is:
  ```xml
  <?xml version="1.0" encoding="UTF-8"?>
  <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
  <plist version="1.0">
    <dict>
      <key>com.apple.security.app-sandbox</key>
      <true/>
      <key>com.apple.security.network.client</key>
      <true/>
      <key>com.apple.security.network.server</key>
      <true/>
      <key>com.apple.security.files.user-selected.read-write</key>
      <true/>
      <key>com.apple.security.files.downloads.read-write</key>
      <true/>
      <key>com.apple.application-identifier</key>
      <string>{{TEAM_ID}}.{{APP_IDENTIFIER}}</string>
      <key>com.apple.developer.team-identifier</key>
      <string>{{TEAM_ID}}</string>
    </dict>
  </plist>
  ```
  > Don't forget to replace TEAM_ID and APP_IDENTIFIER.

**Comment by pronebird** [comment]
It would be great to see PKG added with abilities to:

- Add pre/post install scripts
- Set install-location
- Set min-os-version - you already have it in tauri.conf
- Compression level
- Pkg customizations (pkgbuild) + productbuild with further customizations 
- Pkg notarization

This is what I do so far for outside of AppStore distribution, if that's of any help but it shouldn't be all that different for AppStore distribution:

```sh
# Output dir for the "{APP_NAME}.app" bundle
local app_dir="$APPLE_BUNDLE_DIR/macos"

# Output location for the intermdiate flat package
local pkg_output_path="$APPLE_BUNDLE_DIR/$APPLE_BUNDLE_ID.pkg"

# Output location for the generated package info file
local pkg_info_path="$APPLE_BUNDLE_DIR/PkgInfo";

# Output location for the generated distribution file used by productbuild
local dist_xml_path="$APPLE_BUNDLE_DIR/distribution.xml"

# Generate package info file
pkgbuild --analyze --root "$app_dir" "$pkg_info_path"

# Enforce installation into /Applications
# https://apple.stackexchange.com/a/219144
plutil -replace BundleIsRelocatable -bool NO "$pkg_info_path"

# Allow downgrades, i.e when installing older version over newer.
plutil -replace BundleIsVersionChecked -bool NO "$pkg_info_path"

# Build flat package
pkgbuild --root "$app_dir" \
    --identifier "$APPLE_BUNDLE_ID" \
    --version "$APP_VERSION" \
    --install-location "/Applications" \
    --scripts "$APPLE_PKG_SCRIPTS_DIR" \
    --min-os-version "$APPLE_MIN_OS_VERSION" \
    --compression latest \
    --component-plist "$pkg_info_path" \
    "$pkg_output_path"

# Generate distribution.xml
# https://developer.apple.com/library/content/documentation/DeveloperTools/Reference/DistributionDefinitionRef/Chapters/Distribution_XML_Ref.html
productbuild --synthesize --package "$pkg_output_path" "$dist_xml_path"

# Customize installer:
# - enable_anywhere=false - restrict installation to system volume only, i.e no installs on external drives
# - enable_currentUserHome=false - prevent installation into home dir
# - enable_localSystem=true - enable installation into root
xmlstarlet edit --inplace \
    --subnode '//installer-gui-script' --type elem -n 'domains' \
    --append '//installer-gui-script/domains' --type attr -n 'enable_anywhere' --value 'false' \
    --append '//installer-gui-script/domains' --type attr -n 'enable_currentUserHome' --value 'false' \
    --append '//installer-gui-script/domains' --type attr -n 'enable_localSystem' --value 'true' \
    "$dist_xml_path"

local sign_flags=()
if [[ "$SIGN" == "true" ]]; then
    # codesign with timestamp server because:
    # a. this seems to be a requirement nowadays.
    # b. the certificate can expire but the signature will still be valid.
    sign_flags+=(--sign "$APPLE_INSTALLER_SIGNING_IDENTITY")
    sign_flags+=(--timestamp)
fi

# Customize pkg installer generated by pkgbuild to produce the final installer that will be distributed to users
# See https://stackoverflow.com/a/61551944/351305 for array expansion.
productbuild ${sign_flags[@]+"${sign_flags[@]}"} --distribution "$dist_xml_path" \
                --package-path "$APPLE_BUNDLE_DIR" "$APPLE_DISTRIBUTION_PKG_PATH"

# Remove pkg installer produced by pkgbuild to avoid confusion
rm "$pkg_output_path" || true
```

Notarization part for PKG:

```sh
echo "Notarizing $APPLE_DISTRIBUTION_PKG_PATH"

local notarytool_flags=""
if [[ "$VERBOSE" == "true" ]]; then
    notarytool_flags="--verbose"
fi

# Notarize pkg installer with Apple
xcrun notarytool submit "$APPLE_DISTRIBUTION_PKG_PATH" \
    --apple-id "$APPLE_ID" \
    --team-id "$APPLE_TEAM_ID" \
    --password "$APPLE_PASSWORD" \
    --wait \
    $notarytool_flags

# Attach the notarization ticket to pkg installer
xcrun stapler staple "$APPLE_DISTRIBUTION_PKG_PATH"
```

**Comment by winjeysong** [comment]
> After hours of debugging and research I finally made it work and successfully submitted my app to the app store. Thanks @paprikka and @rendaoer for your inputs.
> 
> The error with testflight was due to missing metadata as pointed out by @paprikka. Specifically
> 
> ```
> <key>DTXcode</key>
> <string>1331</string>
> <key>DTXcodeBuild</key>
> <string>13E500a</string>
> ```
> 
> I had to add those manually, after which testflight accepted my build. This seems like a bug with tauri builder and it should add such fields automatically.
> 
> I also modified the build script to create 2 version of the app - one signed with dev cert and another with dist cert. This will allow to test the signed app locally.
> 
> ```
> targets=("aarch64-apple-darwin" "x86_64-apple-darwin" "universal-apple-darwin");
> 
> target=${targets[$1]};
> 
> app_path="src-tauri/target/${target}/release/bundle/macOS/MyApp.app";
> app_path_dev="src-tauri/target/${target}/release/bundle/macOS/MyApp-dev.app";
> 
> sign_app="Apple Distribution: My Company (XXXXXXXXX)";
> sign_app_dev="Apple Development: My Cert (XXXXXXX)";
> 
> sign_install="3rd Party Mac Developer Installer: My Company (XXXXXXXX)";
> 
> build_name="src-tauri/target/${target}/release/bundle/macOS/MyApp.pkg";
> 
> profile="src-tauri/entitlements/embedded.provisionprofile";
> profile_dev="src-tauri/entitlements/dev/embedded.provisionprofile";
> 
> cp_dir="src-tauri/target/${target}/release/bundle/macOS/MyApp.app/Contents/";
> cp_dir_dev="src-tauri/target/${target}/release/bundle/macOS/MyApp-dev.app/Contents/";
> 
> npm run build:mac;
> cp -r "${app_path}" "${app_path_dev}"
> cp "${profile}" "${cp_dir}";
> cp "${profile_dev}" "${cp_dir_dev}";
> 
> codesign --force -s "${sign_app}" --entitlements ./src-tauri/entitlements/entitlements.mac.plist "${app_path}";
> codesign --force -s "${sign_app_dev}" --entitlements ./src-tauri/entitlements/entitlements.mac.plist "${app_path_dev}";
> 
> productbuild --component "${app_path}" /Applications/ --sign "${sign_install}" "${build_name}";
> ```
> 
> Also if your app is showing a blank window after codesigning you may be missing some entitlements needed by your app. Tauri docs needs to mention clearly that it does not build pkg needed for app store submission and steps needed to build the same. But long term need better support building for app stores for both windows and mac os.

Thanks a lot. It's really a big problem that Tauri dose not include `DTXcode` and `DTXcodeBuild` internally. Without these keys, We cannot submit for TestFlight external Testing:
<img width="658" src="https://github.com/user-attachments/assets/ade1492e-6955-4297-919f-b491602017f0">

**Comment by mkalsi25** [comment]
Hi there,

I am facing this sandbox error while upload on transporter.

<img width="818" alt="Screenshot 2024-12-04 at 5 45 48 PM" src="https://github.com/user-attachments/assets/e8e86e83-c31b-4716-8cc6-5ca0e77b9c2c">

and yes, I have add Entitlement.plist with all required key and value. 
<img width="911" alt="Screenshot 2024-12-04 at 5 50 24 PM" src="https://github.com/user-attachments/assets/17543880-179f-4210-9f8b-d5a8e8dced02">

Is there any other step or important thing we are missing? Please let me know.

**Comment by FabianLars** [comment]
you also need to import that file in tauri.conf.json (for example https://github.com/tauri-apps/tauri/issues/4415#issuecomment-1352041069) - did you do that too?

**Comment by mkalsi25** [comment]
> you also need to import that file in tauri.conf.json (for example [#4415 (comment)](https://github.com/tauri-apps/tauri/issues/4415#issuecomment-1352041069)) - did you do that too?

Yes, I did.  ```macos: {      "entitlements": "../Entitlements.plist", }```

**Comment by wtto00** [comment]
Perhaps missing some keys mentioned in https://github.com/tauri-apps/tauri/issues/4415#issuecomment-1352039657

```xml
    <key>com.apple.application-identifier</key>
    <string>{TEAM_ID}.{BUNDLE_ID}</string>
    
    <key>com.apple.developer.team-identifier</key>
    <string>{TEAM_ID}</string>
```

**Comment by mkalsi25** [comment]
> Perhaps missing some keys mentioned in [#4415 (comment)](https://github.com/tauri-apps/tauri/issues/4415#issuecomment-1352039657)
> 
> ```
>     <key>com.apple.application-identifier</key>
>     <string>{TEAM_ID}.{BUNDLE_ID}</string>
>     
>     <key>com.apple.developer.team-identifier</key>
>     <string>{TEAM_ID}</string>
> ```

do i need to paste it exactly like that? could you elaborate how it gonna help with sandbox issue problem? (apologize for typos*)

**Comment by mkalsi25** [comment]
> Perhaps missing some keys mentioned in [#4415 (comment)](https://github.com/tauri-apps/tauri/issues/4415#issuecomment-1352039657)
> 
> ```
>     <key>com.apple.application-identifier</key>
>     <string>{TEAM_ID}.{BUNDLE_ID}</string>
>     
>     <key>com.apple.developer.team-identifier</key>
>     <string>{TEAM_ID}</string>
> ```

tested! same sandbox issue 🙂‍↕️

**Comment by mkalsi25** [comment]
somehow I fixed my issue by using `signingIdentity` along with `entitlements` in `tauri-conf-json`

```
"entitlements": "../Entitlements.plist",
"signingIdentity": "SIGNING CERT",
```

then use `productbuild` command

Thank you 🙏🏻

**Comment by keypair34** [comment]
> somehow I fixed my issue by using `signingIdentity` along with `entitlements` in `tauri-conf-json`
> 
> ```
> "entitlements": "../Entitlements.plist",
> "signingIdentity": "SIGNING CERT",
> ```
> 
> then use `productbuild` command
> 
> Thank you 🙏🏻

The `signingIdentity` got me.
